internal/handlers/pg/pjson/schema.go

```diff
@@ -0,0 +1,127 @@
+// Copyright 2021 FerretDB Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package pjson
+
+import (
+	"bytes"
+	"encoding/json"
+
+	"github.com/FerretDB/FerretDB/internal/util/lazyerrors"
+)
+
+// schema describes document/object schema needed to unmarshal pjson document.
+type schema struct {
+	Properties map[string]*elem `json:"p"`  // document's properties
```

Comment:
I see that you can't use simple JSON marshal/unmarshal for this case. I want to check if this applies to values of the JSON as well. For example, the values which follows `$s` schema, which will be the next PR. After `$s` we see keys `doc`, `array`, `d`, `s` and `bin`, they are at the same level as `$s`. Can we marshal/unmarshal simply there?

```
   {
    "$s": {
        "$k": ["doc", "array", "d", "s", "bin", "oid", "bool", "t", "null", "regexp", "i32", "ts", "i64"],
        "doc": {
            "t": "object",
            "$s": {
                "$k": ["s"],
                "s": {"t": "string"}
            }
        },
        "array": {
            "t": "array",
            "TODO": "FIXME"
        },
        "d": {"t": "double"},
        "s": {"t": "string"},
        "bin": {"t": "binData", "s": 0},
        "oid": {"t": "objectId"},
        "bool": {"t": "bool"},
        "t": {"t": "date"},
        "null": {"t": "null"},
        "regexp": {"t": "regex","o": "i"},
        "i32": {"t": "int"},
        "ts": {"t": "timestamp"},
        "i64": {"t": "long"}
    },
// here and below
    "doc": {
        "s": "foo"
    },
    "array": [1, 2],
    "d": 42.13,
    "s": "bar",
    "bin": "YmF6",
    "oid": "010101010101010101010101",
    "bool": true,
    "t": -62167219200000,
    "null": null,
    "regex": "hoffman",
    "i32": 42,
    "ts": 1,
    "i64": 9223372036854775807
}
```

Reply:
The problem is not about having or not having `$s` on the same level as values. The problem is having object->array->object->array chains in the schema itself. I tried this approach, which was too complicated (I spent two days and couldn't implement it correctly).

Reply:
Got it, I misunderstood what was complicated. Now got it :pray: